CMakeLists.txt

```diff
@@ -375,17 +374,53 @@ if(LINUX)
   add_definitions(-D__linux__)
 endif(LINUX)
 
-if(ENABLE_SHARED)
-  set(Boost_USE_STATIC_LIBS   OFF)
-else(ENABLE_SHARED)
-  set(Boost_USE_STATIC_LIBS   ON)
-endif(ENABLE_SHARED)
+# Boost
+option(WITH_SYSTEM_BOOST "require and build with system Boost" OFF)
+
+if (WITH_SYSTEM_BOOST)
+  if(ENABLE_SHARED)
+    set(Boost_USE_STATIC_LIBS OFF)
+  else()
+    set(Boost_USE_STATIC_LIBS ON)
+  endif()
+else()
+  set(BOOST_J 1 CACHE STRING "max jobs for Boost build")
```

Comment:
the upside of using `ExternalProject_Add()` is that `cmake` won't take very long to complete, 

- we can only pull the submodules we are interested in using `GIT_REPOSITORY` and `GIT_SUBMODULES`.
- or it can be relatively fast if we just download the dist tarball of boost if we just download it instead of cloning its git repo.
- `cmake` will just do the minimal check and config.

but it might be **very** tedious to add the dependency to boost everywhere we are using it, either header or library.